[pickers] Remove `utils` and `value` params from translations (#14986)
mui/mui-x: docs/data/migration/migration-pickers-v7/migration-pickers-v7.md

:::

Breaking changes that are handled by this codemod are denoted by a ✅ emoji in the table of contents on the right side of the screen.

If you have already applied the `v8.0.0/pickers/preset-safe` (or `v8.0.0/preset-safe`) codemod, then you should not need to take any further action on these items.

All other changes must be handled manually.

:::warning
Not all use cases are covered by codemods. In some scenarios, like props spreading, cross-file dependencies, etc., the changes are not properly identified and therefore must be handled manually.

For example, if a codemod tries to rename a prop, but this prop is hidden with the spread operator, it won't be transformed as expected.

```tsx
<DatePicker {...pickerProps} />
```

After running the codemods, make sure to test your application and that you don't have any console errors.

Feel free to [open an issue](https://github.com/mui/mui-x/issues/new/choose) for support if you need help to proceed with your migration.
:::

> ## New DOM structure for the field

Reviewer: You can remove this section since it's not related to this PR :+1:

```tsx
<DatePicker {...pickerProps} />
```

After running the codemods, make sure to test your application and that you don't have any console errors.

Feel free to [open an issue](https://github.com/mui/mui-x/issues/new/choose) for support if you need help to proceed with your migration.
:::

## Deprecated parameters used on translation keys functions

If you are using a Picker with controlled value and views and you need to compose some of the following translation keys: `openDatePickerDialogue`, `openTimePickerDialogue`, or `clockLabelText`, after upgrading to v8 you only need to pass the formatted value and the time view (only for `clockLabelText`):

```js
const translations = useTranslations();

const openDatePickerDialogue = translations.openDatePickerDialogue(
  value == null ? null : value.format('MM/DD/YYY'),
);
const openTimePickerDialogue = translations.openTimePickerDialogue(
  value == null ? null : value.format('HH:mm:ss'),
);
const clockLabelText = translations.clockLabelText(
  view,
  value == null ? null : value.format('HH:mm:ss'),
);
```

> Also the following types and interfaces no longer receive a generic type parameter:

Reviewer: This breaking change will also be covered by #15001

We can keep it for now waiting for the more detailed section about the removal of `TDate`

  },
});
```

## Deprecated parameters used on translation keys functions

> If you are using a Picker with controlled value and views and you need to compose some of the following translation keys: `openDatePickerDialogue`, `openTimePickerDialogue`, or `clockLabelText`, after upgrading to v8 you only need to pass the formatted value and the time view (only for `clockLabelText`):

Reviewer: > After https://github.com/mui/mui-x/pull/14505 it was possible to call those methods by passing utils as an empty object, wasn't it ?

OK you are right, for the few person (probably only the author of the issue for which we added `formattedValue`) that were using `translations.openTimePickerDialogue` and equivalents, there is a  BC and it can be worth adding the `diff`.

But the main BC remain the other way around :+1:

  },
});
```

## Stop passing `utils` and the date/time object to some translation keys

The following translation keys expect to receive a function: `openDatePickerDialogue`, `openTimePickerDialogue`, or `clockLabelText`. If you are customizing the translation of those keys, you no longer need to provide `utils` and the date/time object to it, but only the formatted date/time value as a string. This affects the customization of those translation keys that on different contexts:

Translations on a single component:

```diff
-<DatePicker localeText={{ openDatePickerDialogue: (date, utils) => string; }} />
+<DatePicker localeText={{ openDatePickerDialogue: (formattedDate) => string; }} />
```

Translations on custom components passed as slots:

```diff
 const translations = useTranslations();
-const openDatePickerDialogue = translations.openTimePickerDialogue(value, {} as any, value == null ? null : value.format('MM/DD/YYY'));
+const openDatePickerDialogue = translations.openDatePickerDialogue(value == null ? null : value.format('MM/DD/YYY'));
```

> Notice that `clockLabelText` expects an additional parameter with the time view.

Reviewer: I would suggest just doing on diff per translation key.
The migration guide can be very lengthy if it helps people migrate (see [this section](https://mui.com/x/migration/migration-pickers-v6/#adapters-internal-changes) for a very long migration content :laughing: )

So something like:

````
The following translation key no longer receive the `utils` and the date object:

- `openDatePickerDialogue`:

  ```diff
  // If you are setting a custom value for this translation key in a picker component or in the `LocalizationProvider`
  -<DatePicker localeText={{ openDatePickerDialogue: (date, utils) => string; }} />
  +<DatePicker localeText={{ openDatePickerDialogue: (formattedDate) => string; }} />

  // If you using this translation key in a custom component
   const translations = useTranslations();
  -const openDatePickerDialogue = translations.openTimePickerDialogue(value, {} as any, value == null ? null : value.format('MM/DD/YYY'));
  +const openDatePickerDialogue = translations.openDatePickerDialogue(value == null ? null : value.format('MM/DD/YYY'));
  ```

  - `openTimePickerDialogue`:

  ```diff
  ```
````

  },
});
```

> ## Stop passing `utils` and the date/time object to some translation keys

Reviewer 2: Technically there is no such thing as a `time` object. 🙈 🤷 
```suggestion
## Stop passing `utils` and the date object to some translation keys
```

> If you are customizing the translation of those keys, you no longer need to provide `utils` and the date/time object to it, but only the formatted date/time value as a string.

Reviewer 2: Grammarly suggestion. 😉 
```suggestion
If you are customizing the translation of those keys, you no longer need to provide `utils` and the date object, but only the formatted date value as a string.
```

Reviewer 2: Oh, sorry for the confusion.
The removal of "to it" was suggested by Grammarly, but the `/time` removal was from me. 🙈 
I think that we should use Date and Time Pickers if we want to refer to the components,
In this context it feels like `date/time` refers to the date object mainly. 🤷

> When migrating to v8, you must have a small refactor on the usage of those keys:

Reviewer 2: ```suggestion
If you have customized the following translation keys, you have to update them:
```

- `openDatePickerDialogue`:

>   ```diff

Reviewer 2: Have you considered combining these diffs into one?
I feel like it would create less visual noise. 🤔

  +const openDatePickerDialogue = translations.openDatePickerDialogue(value == null ? null : value.format('MM/DD/YYY'));
  ```

  - `openTimePickerDialogue`:

  ```diff
   // If you are setting a custom value in a picker component or in the `LocalizationProvider`
  -<DatePicker localeText={{ openTimePickerDialogue: (time, utils) => string; }} />
  +<DatePicker localeText={{ openTimePickerDialogue: (formattedTime) => string; }} />

   // If you are setting a custom value in the `LocalizationProvider`
  -<LocalizationProvider localeText={{ openTimePickerDialogue: (time, utils) => string; }} >
  +<LocalizationProvider localeText={{ openTimePickerDialogue: (formattedTime) => string; }} >

   // If you using this translation key in a custom component
   const translations = useTranslations();
  -const openTimePickerDialogue = translations.openTimePickerDialogue(value, {} as any, value == null ? null : value.format('hh:mm:ss'));
  +const openTimePickerDialogue = translations.openTimePickerDialogue(value == null ? null : value.format('hh:mm:ss'));
  ```

>   - `clockLabelText`:

Reviewer 2: WDYT about putting it as the first item in alphabetical order?

  },
});
```

## Stop passing `utils` and the date object to some translation keys

Some translation keys no longer require `utils` and the date object, but only the formatted date value as a string.

If you have customized the following translation keys, you have to update them:

```diff
 /**
 * If you are setting a custom value in a picker component
 */

 //clockLabelText

> -<DatePicker localeText={{ clockLabelText: (view, time, utils) => string; }} />

Reviewer 2: Great observations. 👍

+<DatePicker localeText={{ clockLabelText: (view, formattedTime) => string; }} />

 // openDatePickerDialogue
-<DatePicker localeText={{ openDatePickerDialogue: (date, utils) => string; }} />
+<DatePicker localeText={{ openDatePickerDialogue: (formattedDate) => string; }} />

 // openTimePickerDialogue
-<DatePicker localeText={{ openTimePickerDialogue: (time, utils) => string; }} />
+<DatePicker localeText={{ openTimePickerDialogue: (formattedTime) => string; }} />

 /**
 * If you are setting a custom value in the `LocalizationProvider`
 */

 //clockLabelText

> -<LocalizationProvider localeText={{ clockLabelText: (view, time, utils) => string; }} >

Reviewer 2: Nice, that's probably the best approach. 💯

  > &
    UseDateTimeFieldProps<TDate, TEnableAccessibleFieldDOMStructure>;
  ```

## Stop passing `utils` and the date object to some translation keys

Some translation keys no longer require `utils` and the date object as parameters, but only the formatted value as a string. The keys affected by this changes are: `clockLabelText`, `openDatePickerDialogue` and `openTimePickerDialogue`.

If you have customized those translation keys, you have to update them:

> ```diff

Reviewer 2: Given the increased amount of examples, I no longer strongly prefer grouping them in a single diff. 😉 
If you feel like 3 separate diffs per type of usage make more sense, I'm all for it. 👍

 */
 <LocalizationProvider localeText={{
-   clockLabelText: (view, time, utils) =>
-     `Select ${view}. ${
-       time === null || !utils.isValid(time)
-         ? 'No time selected'
-         : `Selected time is ${utils.format(time, 'fullTime')}`
-     }`
+   clockLabelText: (view, formattedTime) =>
+     `Select ${view}. ${
+       formattedTime == null ? 'No time selected' : `Selected time is ${formattedTime}`
+     }`
-   openDatePickerDialogue: (value, utils) =>
-     value !== null && utils.isValid(value)
-      ? `Choose date, selected date is ${utils.format(value, 'fullDate')}`
-      : 'Choose date',
+   openDatePickerDialogue: (formattedDate) =>
+     formattedDate ? `Choose date, selected date is ${formattedDate}` : 'Choose date'
-   openTimePickerDialogue: (value, utils) =>
-     value !== null && utils.isValid(value)
-       ? `Choose time, selected time is ${utils.format(value, 'fullTime')}`
-       : 'Choose time',
+   openTimePickerDialogue: (formattedTime) =>
+     formattedTime ? `Choose time, selected time is ${formattedTime}` : 'Choose time'
 }} >

 /**
 * If you using this translation key in a custom component
 */

>  const translations = useTranslations();

Reviewer 2: ```suggestion
 const translations = usePickersTranslations();
```
